Security Incident Disclosure (#632)
Homebrew/brew.sh: _posts/2021-04-21-security-incident-disclosure.md

---
title: Security Incident Disclosure
author: reitermarkus
---

> On 18th April 2021 [a security researcher identified](https://blog.ryotak.me/post/homebrew-security-incident-en/) a vulnerability in our `review-cask-pr` GitHub Action used to automatically review and approve cask version bump pull requests which was automatically merged by our `automerge` GitHub Action. This was reported via HackerOne and [a pull request demonstrating the vulnerability](https://github.com/Homebrew/homebrew-cask/pull/104191) was submitted with our permission. The [pull request was subsequently reverted](https://github.com/Homebrew/homebrew-cask/pull/104197), the [`automerge` action disabled and removed](https://github.com/Homebrew/homebrew-cask/pull/104199) and the [`review-cask-pr` action disabled and removed](https://github.com/Homebrew/homebrew-cask/pull/104204) from all the `homebrew-cask*` repositories that were using it.

Reviewer: ```suggestion
On 18th April 2021, [a security researcher identified](https://blog.ryotak.me/post/homebrew-security-incident-en/) a vulnerability in our `review-cask-pr` GitHub Action. The vulnerability affected:

1. the `homebrew-cask` repository and
2. all `homebrew-cask-*` repositories (taps) on the Homebrew organization, except those for which we never enabled the Action.

Whenever an affected cask repository received a pull request which just changed the version of a cask and nothing else, the `review-cask-pr` GitHub Action would automatically review and approve the pull request. The approval would then trigger another GitHub Action named `automerge`, which in turn would pick up the approved pull request and merge it. This was reported via HackerOne and [a pull request demonstrating the vulnerability](https://github.com/Homebrew/homebrew-cask/pull/104191) (PoC) was submitted with our permission. We subsequently [reverted the PoC pull request](https://github.com/Homebrew/homebrew-cask/pull/104197), [disabled and removed the `automerge` Action](https://github.com/Homebrew/homebrew-cask/pull/104199) and [disabled and removed the `review-cask-pr` Action](https://github.com/Homebrew/homebrew-cask/pull/104204) from the affected cask repositories.
```

Summary of my changes:

1. use simpler language to make the text easier to read;
2. clarify that the Actions were written by us, not by GitHub;
3. clarify that `homebrew-cask` was specifically affected;
4. clarify that a number of taps were not affected (I don’t know which ones though);
5. clarify what a cask version bump is;
6. clarify which one of the pull requests mentioned was reverted (i. e., the PoC one);
7. use the active voice more often;
8. insert a missing comma;
9. capitalize the word _Action_ consistently.

Reviewer 2: Sounds good, a few quibbles:

> * clarify that a number of taps were not affected (I don’t know which ones though);

All homebrew-cask* taps were technically affected by the vulnerability and brew/homebrew-core/external command taps were not.

> 9\. capitalize the word _Action_ consistently.

I think we should use "GitHub Action" or "GitHub Actions" but "action" or "actions" (without the GitHub prefix).

> 6\. (i. e., the PoC one);

Maybe worth making this "proof of concept" for those unfamiliar? 

@claui any chance you could amend this accordingly?

Reviewer 2: Actually, ignore me, I'll apply this then edit it.

### What was impacted

> The discovered vulnerability would allow an attacker to inject arbitrary code into a cask and have it be merged automatically. This is due to a flaw in the `git_diff` dependency of the `review-cask-pr` GitHub Action which is used to parse a pull request's diff for inspection. Due to this flaw, the parser can be spoofed into completely ignoring the offending lines, resulting in successfully approving a malicious pull request.

Reviewer: ```suggestion
The discovered vulnerability would allow an attacker to inject arbitrary code into a cask and have it be merged automatically. This is due to a flaw in the `git_diff` dependency of the `review-cask-pr` GitHub Action, which is used to parse a pull request's diff for inspection. Due to this flaw, the parser can be spoofed into completely ignoring the offending lines, resulting in successfully approving a malicious pull request.
```

Just adding a comma here.

**A single cask was compromised with a harmless change for the duration of [the demonstration pull request](https://github.com/Homebrew/homebrew-cask/pull/104191) until [its reversal](https://github.com/Homebrew/homebrew-cask/pull/104197). No action is required by users due to this incident.**

### What we're doing about it

- The vulnerable `review-cask-pr` GitHub Action has been disabled and removed from all repositories.
- The `automerge` GitHub Action has been disabled and removed from all repositories (in favour of the GitHub built-in functionality that did not exist when this action was created).
- We have removed the ability for our bots to commit to `homebrew/cask*` repositories.
- All `homebrew/cask*` pull requests will require a manual review and approval by a maintainer.
- We are improving documentation to help onboard new `homebrew/cask` maintainers and training

>   existing `homebrew/core` maintainers to help with `homebrew/core`.

Reviewer 3: ```suggestion
  existing `homebrew/core` maintainers to help with `homebrew/cask`.
```
If I understood it correctly

---
title: Security Incident Disclosure
author: reitermarkus
---

On 18th April 2021, [a security researcher identified](https://blog.ryotak.me/post/homebrew-security-incident-en/) a vulnerability in our `review-cask-pr` GitHub Action used on the `homebrew-cask` and all `homebrew-cask-*` taps (non-default repositories) in the Homebrew organization.

> Whenever an affected cask tap received a pull request to change the only version of a cask, the `review-cask-pr` GitHub Action would automatically review and approve the pull request. The approval would then trigger the `automerge` GitHub Action which would merge the approved pull request. This was reported via HackerOne and [a proof-of-concept (PoC) pull request demonstrating the vulnerability](https://github.com/Homebrew/homebrew-cask/pull/104191) was submitted with our permission. We subsequently [reverted the PoC pull request](https://github.com/Homebrew/homebrew-cask/pull/104197), [disabled and removed the `automerge` GitHub Action](https://github.com/Homebrew/homebrew-cask/pull/104199) and [disabled and removed the `review-cask-pr` GitHub Action](https://github.com/Homebrew/homebrew-cask/pull/104204) from all vulnerable repositories.

Reviewer 4: ```
Whenever an affected cask tap received a pull request to change the only version of a cask, the `review-cask-pr` GitHub Action would automatically review and approve the pull request. The approval would then trigger the `automerge` GitHub Action which would merge the approved pull request. This was reported via HackerOne 
```

Isn't this `this` should mention the `a vulnerability in our `review-cask-pr` GitHub Action`?

> On 18th April 2021, [a security researcher identified](https://blog.ryotak.me/post/homebrew-security-incident-en/) a vulnerability in our `review-cask-pr` GitHub Action used on the `homebrew-cask` and all `homebrew-cask-*` taps (non-default repositories) in the Homebrew organization.
> 
> Whenever an affected cask tap received a pull request to change the only version of a cask, the `review-cask-pr` GitHub Action would automatically review and approve the pull request. The approval would then trigger the `automerge` GitHub Action which would merge the approved pull request. This was reported via HackerOne and [a proof-of-concept (PoC) pull request demonstrating the vulnerability](https://github.com/Homebrew/homebrew-cask/pull/104191) was submitted with our permission. We subsequently [reverted the PoC pull request](https://github.com/Homebrew/homebrew-cask/pull/104197), [disabled and removed the `automerge` GitHub Action](https://github.com/Homebrew/homebrew-cask/pull/104199) and [disabled and removed the `review-cask-pr` GitHub Action](https://github.com/Homebrew/homebrew-cask/pull/104204) from all vulnerable repositories.

Reviewer 2: ```suggestion
On 18th April 2021, [a security researcher identified](https://blog.ryotak.me/post/homebrew-security-incident-en/) a vulnerability in our `review-cask-pr` GitHub Action used on the `homebrew-cask` and all `homebrew-cask-*` taps (non-default repositories) in the Homebrew organization and [reported it on our HackerOne](https://hackerone.com/homebrew).

Whenever an affected cask tap received a pull request to change the only version of a cask, the `review-cask-pr` GitHub Action would automatically review and approve the pull request. The approval would then trigger the `automerge` GitHub Action which would merge the approved pull request. [A proof-of-concept (PoC) pull request demonstrating the vulnerability](https://github.com/Homebrew/homebrew-cask/pull/104191) was submitted with our permission. We subsequently [reverted the PoC pull request](https://github.com/Homebrew/homebrew-cask/pull/104197), [disabled and removed the `automerge` GitHub Action](https://github.com/Homebrew/homebrew-cask/pull/104199) and [disabled and removed the `review-cask-pr` GitHub Action](https://github.com/Homebrew/homebrew-cask/pull/104204) from all vulnerable repositories.
```

> Whenever an affected cask tap received a pull request to change the only version of a cask, the `review-cask-pr` GitHub Action would automatically review and approve the pull request. The approval would then trigger the `automerge` GitHub Action which would merge the approved pull request. [A proof-of-concept (PoC) pull request demonstrating the vulnerability](https://github.com/Homebrew/homebrew-cask/pull/104191) was submitted with our permission. We subsequently [reverted the PoC pull request](https://github.com/Homebrew/homebrew-cask/pull/104197), [disabled and removed the `automerge` GitHub Action](https://github.com/Homebrew/homebrew-cask/pull/104199) and [disabled and removed the `review-cask-pr` GitHub Action](https://github.com/Homebrew/homebrew-cask/pull/104204) from all vulnerable repositories.

Reviewer: ```suggestion
Whenever an affected cask tap received a pull request to change only the version of a cask, the `review-cask-pr` GitHub Action would automatically review and approve the pull request. The approval would then trigger the `automerge` GitHub Action which would merge the approved pull request. [A proof-of-concept (PoC) pull request demonstrating the vulnerability](https://github.com/Homebrew/homebrew-cask/pull/104191) was submitted with our permission. We subsequently [reverted the PoC pull request](https://github.com/Homebrew/homebrew-cask/pull/104197), [disabled and removed the `automerge` GitHub Action](https://github.com/Homebrew/homebrew-cask/pull/104199) and [disabled and removed the `review-cask-pr` GitHub Action](https://github.com/Homebrew/homebrew-cask/pull/104204) from all vulnerable repositories.
```

---
title: Security Incident Disclosure
author: reitermarkus
---

> On 18th April 2021, [a security researcher identified](https://blog.ryotak.me/post/homebrew-security-incident-en/) a vulnerability in our `review-cask-pr` GitHub Action. The vulnerability affected:
> 
> 1. the `homebrew-cask` repository and
> 2. all `homebrew-cask-*` repositories (taps) on the Homebrew organization, except those for which we never enabled the Action.
> 
> Whenever an affected cask repository received a pull request which just changed the version of a cask and nothing else, the `review-cask-pr` GitHub Action would automatically review and approve the pull request. The approval would then trigger another GitHub Action named `automerge`, which in turn would pick up the approved pull request and merge it. This was reported via HackerOne and [a pull request demonstrating the vulnerability](https://github.com/Homebrew/homebrew-cask/pull/104191) (PoC) was submitted with our permission. We subsequently [reverted the PoC pull request](https://github.com/Homebrew/homebrew-cask/pull/104197), [disabled and removed the `automerge` Action](https://github.com/Homebrew/homebrew-cask/pull/104199) and [disabled and removed the `review-cask-pr` Action](https://github.com/Homebrew/homebrew-cask/pull/104204) from the affected cask repositories.
> 

Reviewer 2: ```suggestion
On 18th April 2021, [a security researcher identified](https://blog.ryotak.me/post/homebrew-security-incident-en/) a vulnerability in our `review-cask-pr` GitHub Action used on the `homebrew-cask` and all `homebrew-cask-*` taps (non-default repositories) in the Homebrew organization.

Whenever an affected cask tap received a pull request to change the only version of a cask, the `review-cask-pr` GitHub Action would automatically review and approve the pull request. The approval would then trigger the `automerge` GitHub Action which would merge the approved pull request. This was reported via HackerOne and [a proof-of-concept (PoC) pull request demonstrating the vulnerability](https://github.com/Homebrew/homebrew-cask/pull/104191) was submitted with our permission. We subsequently [reverted the PoC pull request](https://github.com/Homebrew/homebrew-cask/pull/104197), [disabled and removed the `automerge` GitHub Action](https://github.com/Homebrew/homebrew-cask/pull/104199) and [disabled and removed the `review-cask-pr` GitHub Action](https://github.com/Homebrew/homebrew-cask/pull/104204) from all vulnerable repositories.

```
